File: java-sdk/src/conch/plugin/HostApi.java
package conch.plugin;

/**
 * Host API for Conch plugins.
 *
 * <p>Provides static methods that plugins call to interact with the Conch
 * terminal emulator. All methods are backed by native (Rust/JNI)
 * implementations that are registered by the host before any plugin code
 * runs.</p>
 *
 * <p>These methods are <b>thread-safe</b> and can be called from any thread,
 * though they are typically called from the plugin's dedicated thread.</p>
 *
 * <h2>Available Operations</h2>
 * <ul>
 *   <li><b>Logging</b> — {@link #log}, {@link #info}, {@link #warn},
 *       {@link #error}, {@link #debug}, {@link #trace}</li>
 *   <li><b>Menu items</b> — {@link #registerMenuItem}</li>
 * </ul>
 *
 * <h2>Example</h2>
 * <pre>{@code
 * public void setup() {
 *     HostApi.info("Plugin starting up...");
 *     HostApi.registerMenuItem("Tools", "Run Analysis", "run_analysis");
 * }
 *
 * public void onEvent(String eventJson) {
 *     if (eventJson.contains("run_analysis")) {
 *         HostApi.info("Running analysis...");
 *         // ... do work ...
 *         HostApi.info("Analysis complete.");
 *     }
 * }
 * }</pre>
 *
 * @see ConchPlugin
 */
public class HostApi {
    private HostApi() {} // Static-only class — not instantiable.

    // -----------------------------------------------------------------------
    // Logging
    // -----------------------------------------------------------------------

    /** Log level constant: trace (most verbose). */
    public static final int LOG_TRACE = 0;
    /** Log level constant: debug. */
    public static final int LOG_DEBUG = 1;
    /** Log level constant: info. */
    public static final int LOG_INFO = 2;
    /** Log level constant: warn. */
    public static final int LOG_WARN = 3;
    /** Log level constant: error (least verbose). */
    public static final int LOG_ERROR = 4;

    /**
     * Log a message at the specified level.
     *
     * <p>Messages appear in Conch's log output (visible when running with
     * {@code RUST_LOG=info} or similar).</p>
     *
     * @param level one of {@link #LOG_TRACE}, {@link #LOG_DEBUG},
     *              {@link #LOG_INFO}, {@link #LOG_WARN}, {@link #LOG_ERROR}
     * @param message the message to log (must not be null)
     */
    public static native void log(int level, String message);

    /**
     * Log a message at TRACE level.
     *
     * @param message the message to log
     */
    public static void trace(String message) { log(LOG_TRACE, message); }

    /**
     * Log a message at DEBUG level.
     *
     * @param message the message to log
     */
    public static void debug(String message) { log(LOG_DEBUG, message); }

    /**
     * Log a message at INFO level.
     *
     * @param message the message to log
     */
    public static void info(String message) { log(LOG_INFO, message); }

    /**
     * Log a message at WARN level.
     *
     * @param message the message to log
     */
    public static void warn(String message) { log(LOG_WARN, message); }

    /**
     * Log a message at ERROR level.
     *
     * @param message the message to log
     */
    public static void error(String message) { log(LOG_ERROR, message); }

    // -----------------------------------------------------------------------
    // Menu Items
    // -----------------------------------------------------------------------

    /**
     * Register a menu item in Conch's menu bar.
     *
     * <p>The item appears under the specified top-level menu. When the user
     * clicks it, the plugin's {@link ConchPlugin#onEvent(String)} is called
     * with a JSON event containing the {@code action} string:</p>
     *
     * <pre>{@code {"kind":"menu_action","action":"your_action_id"}}</pre>
     *
     * <p>Typically called during {@link ConchPlugin#setup()}.</p>
     *
     * @param menu   the top-level menu name (e.g. {@code "Tools"})
     * @param label  the menu item label shown to the user
     * @param action the action identifier included in the event when clicked
     */
    public static native void registerMenuItem(String menu, String label, String action);

    /**
     * Register a menu item with a keyboard shortcut.
     *
     * @param menu    the top-level menu name
     * @param label   the menu item label
     * @param action  the action identifier
     * @param keybind keyboard shortcut (e.g. {@code "cmd+shift+j"})
     */
    public static native void registerMenuItemWithKeybind(String menu, String label, String action, String keybind);

    // -----------------------------------------------------------------------
    // Notifications
    // -----------------------------------------------------------------------

    /**
     * Show a toast notification.
     *
     * @param title      notification title (may be null)
     * @param body       notification body text
     * @param level      one of {@code "info"}, {@code "success"},
     *                   {@code "warn"}, {@code "error"}
     * @param durationMs display duration in milliseconds (0 = persistent,
     *                   -1 = default 5 seconds)
     */
    public static native void notify(String title, String body, String level, int durationMs);

    /**
     * Show a toast notification with default duration.
     *
     * @param title notification title
     * @param body  notification body text
     * @param level notification level
     */
    public static void notify(String title, String body, String level) {
        notify(title, body, level, -1);
    }

    // -----------------------------------------------------------------------
    // Status Bar
    // -----------------------------------------------------------------------

    /**
     * Update the global status bar.
     *
     * @param text     status text to display (null to clear)
     * @param level    0=info, 1=warn, 2=error, 3=success
     * @param progress 0.0–1.0 to show a progress bar, or negative to hide it
     */
    public static native void setStatus(String text, int level, float progress);

    // -----------------------------------------------------------------------
    // Clipboard
    // -----------------------------------------------------------------------

    /**
     * Set the system clipboard contents.
     *
     * @param text the text to copy to the clipboard
     */
    public static native void clipboardSet(String text);

    /**
     * Get the system clipboard contents.
     *
     * @return the clipboard text, or null if unavailable
     */
    public static native String clipboardGet();

    // -----------------------------------------------------------------------
    // Plugin Config (persistent key/value storage)
    // -----------------------------------------------------------------------

    /**
     * Read a persisted config value for this plugin.
     *
     * <p>Config is stored per-plugin in
     * {@code ~/.config/conch/plugins/<plugin-name>/<key>.json}.</p>
     *
     * @param key the config key
     * @return the JSON value string, or null if not set
     */
    public static native String getConfig(String key);

    /**
     * Write a persisted config value for this plugin.
     *
     * @param key   the config key
     * @param value the JSON value string (null to delete)
     */
    public static native void setConfig(String key, String value);

    // -----------------------------------------------------------------------
    // Dialogs
    // -----------------------------------------------------------------------

    /**
     * Show a blocking text prompt dialog and wait for user input.
     *
     * <p>The dialog blocks the plugin thread until the user submits or
     * cancels. Returns the entered text, or null if cancelled.</p>
     *
     * @param message      the prompt message
     * @param defaultValue pre-filled default value (use "" for empty)
     * @return the entered text, or null if cancelled
     */
    public static native String prompt(String message, String defaultValue);

    /**
     * Show a blocking text prompt with no default value.
     *
     * @param message the prompt message
     * @return the entered text, or null if cancelled
     */
    public static String prompt(String message) {
        return prompt(message, "");
    }

    /**
     * Show a blocking confirmation dialog (OK / Cancel).
     *
     * @param message the confirmation message
     * @return true if the user confirmed, false if cancelled
     */
    public static native boolean confirm(String message);

    /**
     * Show a blocking alert dialog with an OK button.
     *
     * @param title   the alert title
     * @param message the alert message
     */
    public static native void alert(String title, String message);

    /**
     * Show a blocking error dialog with an OK button.
     *
     * @param title   the error title
     * @param message the error message
     */
    public static native void showError(String title, String message);

    // -----------------------------------------------------------------------
    // Forms
    // -----------------------------------------------------------------------

    /**
     * Show a blocking multi-field form dialog.
     *
     * <p>The form descriptor is a JSON string describing the title and fields.
     * Supported field types: {@code text}, {@code password}, {@code number},
     * {@code combo} (dropdown), {@code checkbox}, {@code host_port},
     * {@code file_picker}, {@code collapsible}, {@code separator},
     * {@code label}.</p>
     *
     * <p><b>Example:</b></p>
     * <pre>{@code
     * String formJson = """
     *     {
     *         "title": "New Connection",
     *         "fields": [
     *             {"type": "text", "id": "host", "label": "Hostname", "hint": "e.g. server.example.com"},
     *             {"type": "number", "id": "port", "label": "Port", "value": 22},
     *             {"type": "text", "id": "user", "label": "Username"},
     *             {"type": "password", "id": "pass", "label": "Password"},
     *             {"type": "combo", "id": "auth", "label": "Auth Method",
     *              "options": ["password", "key", "agent"], "value": "password"},
     *             {"type": "checkbox", "id": "save", "label": "Save credentials", "value": true}
     *         ]
     *     }
     *     """;
     * String result = HostApi.showForm(formJson);
     * if (result != null) {
     *     // result is a JSON object: {"host":"...", "port":22, "user":"...", ...}
     *     JsonObject obj = JsonParser.parseString(result).getAsJsonObject();
     *     String host = obj.get("host").getAsString();
     * }
     * }</pre>
     *
     * @param formDescriptorJson JSON string with {@code title} and {@code fields} array
     * @return JSON object with field values keyed by id, or null if cancelled
     */
    public static native String showForm(String formDescriptorJson);

    // -----------------------------------------------------------------------
    // Inter-Plugin Communication
    // -----------------------------------------------------------------------

    /**
     * Subscribe to bus events from other plugins.
     *
     * <p>When a matching event is published, the plugin's
     * {@link ConchPlugin#onEvent(String)} receives it as a
     * {@code bus_event} JSON envelope.</p>
     *
     * @param eventType the event type to subscribe to (e.g. {@code "ssh.connected"})
     */
    public static native void subscribe(String eventType);

    /**
     * Publish an event on the plugin bus.
     *
     * @param eventType the event type string
     * @param dataJson  JSON-encoded event data
     */
    public static native void publishEvent(String eventType, String dataJson);

    // -----------------------------------------------------------------------
    // Terminal / Session
    // -----------------------------------------------------------------------

    /**
     * Write text to the focused window's active terminal session.
     *
     * <p>The text is sent as raw bytes to the PTY. Include {@code "\n"} to
     * simulate pressing Enter.</p>
     *
     * @param text the text to write (e.g. {@code "ls -la\n"})
     */
    public static native void writeToPty(String text);

    /**
     * Open a new local shell tab in the focused window.
     *
     * @param command optional command to run in the new tab (null for none)
     * @param plain   if true, use the OS default shell instead of the
     *                configured terminal.shell (avoids nesting tmux, etc.)
     */
    public static native void newTab(String command, boolean plain);

    /**
     * Open a new tab with default shell configuration.
     */
    public static void newTab() {
        newTab(null, false);
    }

    /**
     * Open a new plain shell tab and run a command.
     *
     * @param command the command to execute in the new tab
     */
    public static void newPlainTab(String command) {
        newTab(command, true);
    }
}

File: crates/conch_plugin/src/host_api.rs
//! Safe Rust host API trait for Lua/Java/Tauri plugin integration.
//!
//! Each plugin gets its own `Arc<dyn HostApi>` instance with the plugin name
//! baked in, eliminating the fragile thread-name-based identification.

use conch_plugin_sdk::PanelLocation;

/// Safe Rust interface for the host API.
///
/// Implemented by `TauriHostApi` in `conch_tauri`.
pub trait HostApi: Send + Sync {
    fn plugin_name(&self) -> &str;
    fn check_permission(&self, _capability: &str) -> bool {
        true
    }

    // -- Panel Management --
    fn register_panel(&self, location: PanelLocation, name: &str, icon: Option<&str>) -> u64;
    fn set_widgets(&self, handle: u64, widgets_json: &str);

    // -- Logging & Notifications --
    fn log(&self, level: u8, msg: &str);
    fn notify(&self, json: &str);
    fn set_status(&self, text: Option<&str>, level: u8, progress: f32);

    // -- Event Bus --
    fn publish_event(&self, event_type: &str, data_json: &str);
    fn subscribe(&self, event_type: &str);
    fn query_plugin(&self, target: &str, method: &str, args_json: &str) -> Option<String>;
    fn register_service(&self, name: &str);

    // -- Config Persistence --
    fn get_config(&self, key: &str) -> Option<String>;
    fn set_config(&self, key: &str, value: &str);

    // -- Clipboard --
    fn clipboard_set(&self, text: &str);
    fn clipboard_get(&self) -> Option<String>;

    // -- Theme --
    fn get_theme(&self) -> Option<String>;

    // -- Menu --
    fn register_menu_item(&self, menu: &str, label: &str, action: &str, keybind: Option<&str>);

    // -- Dialogs (blocking — called from plugin thread) --
    fn show_form(&self, json: &str) -> Option<String>;
    fn show_confirm(&self, msg: &str) -> bool;
    fn show_prompt(&self, msg: &str, default_value: &str) -> Option<String>;
    fn show_alert(&self, title: &str, msg: &str);
    fn show_error(&self, title: &str, msg: &str);
    fn show_context_menu(&self, json: &str) -> Option<String>;

    // -- Terminal / Tabs --
    fn write_to_pty(&self, data: &[u8]);
    fn new_tab(&self, command: Option<&str>, plain: bool);

    // -- Session Management --
    fn open_session(&self, meta_json: &str) -> u64;
    fn close_session(&self, handle: u64);
    fn set_session_status(&self, handle: u64, status: u8, detail: Option<&str>);
    fn session_prompt(
        &self,
        handle: u64,
        prompt_type: u8,
        msg: &str,
        detail: Option<&str>,
    ) -> Option<String>;
}

#[cfg(test)]
mod tests {
    use super::*;

    struct MockHostApi {
        name: String,
    }

    impl HostApi for MockHostApi {
        fn plugin_name(&self) -> &str {
            &self.name
        }
        fn register_panel(&self, _: PanelLocation, _: &str, _: Option<&str>) -> u64 {
            1
        }
        fn set_widgets(&self, _: u64, _: &str) {}
        fn log(&self, _: u8, _: &str) {}
        fn notify(&self, _: &str) {}
        fn set_status(&self, _: Option<&str>, _: u8, _: f32) {}
        fn publish_event(&self, _: &str, _: &str) {}
        fn subscribe(&self, _: &str) {}
        fn query_plugin(&self, _: &str, _: &str, _: &str) -> Option<String> {
            None
        }
        fn register_service(&self, _: &str) {}
        fn get_config(&self, _: &str) -> Option<String> {
            None
        }
        fn set_config(&self, _: &str, _: &str) {}
        fn clipboard_set(&self, _: &str) {}
        fn clipboard_get(&self) -> Option<String> {
            None
        }
        fn get_theme(&self) -> Option<String> {
            None
        }
        fn register_menu_item(&self, _: &str, _: &str, _: &str, _: Option<&str>) {}
        fn show_form(&self, _: &str) -> Option<String> {
            None
        }
        fn show_confirm(&self, _: &str) -> bool {
            false
        }
        fn show_prompt(&self, _: &str, _: &str) -> Option<String> {
            None
        }
        fn show_alert(&self, _: &str, _: &str) {}
        fn show_error(&self, _: &str, _: &str) {}
        fn show_context_menu(&self, _: &str) -> Option<String> {
            None
        }
        fn write_to_pty(&self, _: &[u8]) {}
        fn new_tab(&self, _: Option<&str>, _: bool) {}
        fn open_session(&self, _: &str) -> u64 {
            0
        }
        fn close_session(&self, _: u64) {}
        fn set_session_status(&self, _: u64, _: u8, _: Option<&str>) {}
        fn session_prompt(&self, _: u64, _: u8, _: &str, _: Option<&str>) -> Option<String> {
            None
        }
    }

    #[test]
    fn mock_host_api_implements_trait() {
        let api: Box<dyn HostApi> = Box::new(MockHostApi {
            name: "test".into(),
        });
        assert_eq!(api.plugin_name(), "test");
        assert!(!api.show_confirm("?"));
    }

    #[test]
    fn trait_object_is_send_sync() {
        fn assert_send_sync<T: Send + Sync>() {}
        assert_send_sync::<Box<dyn HostApi>>();
    }
}
